src/tools/non-promotion-reasons.md

```diff
@@ -111,16 +163,546 @@ one or more of the following:
+test(C c) {
+  if (c.n != null) {
+    print(c.n + 1); // ERROR
+  }
+}
+```
+
+Here, the compiler produces an error message
+explaining that the `+` operator can't be called on `n`
+because it is potentially null.
+
+**Message:**
+
+```nocode
+'n' refers to a public field so it couldn’t be promoted.
+```
+
+**Solution:**
+
+Making the field private lets the compiler be sure that no outside libraries
+could possibly override its value, so it's safe to promote.
+
+{:.good}
+```dart
+class C {
+  final int? _n;
+  C(this.n);
+}
+
+test(C c) {
+  if (c.n != null) {
+    print(c.n + 1); // OK
+  }
+}
+```
+
+### Only final fields can be promoted {#final}
+
+**The cause:**
+You're trying to promote a field, but the field is not final.
+
+To the compiler, non-final fields could, in principle,
+be modified any time between the time
+they’re tested and the time they’re used.
+So it's not safe for the compiler to promote a non-final nullable type
+to a non-nullable type.
+
+**Example:**
+
+{:.bad}
+```dart
+class C {
+  int? _mutablePrivateField; // CONTEXT
+  Example(this._mutablePrivateField);
+
+  f() {
+    if (_mutablePrivateField != null) {
+      int i = _mutablePrivateField; // ERROR
+    }
+  }
+}
+```
+
+**Message:**
+
+```nocode
+'mutablePrivateField' refers to a non-final field so it couldn’t be promoted.
+```
+
+**Solution:**
+
+Make the field `final`
+
+{:.good}
+```dart
+class Example {
+  int? final _immutablePrivateField; 
+  Example(this._immutablePrivateField);
+
+  f() {
+    if (_mutablePrivateField != null) {
+      int i = _immutablePrivateField; // OK
+    }
+  }
+}
+```
+
+### Getters can't be promoted {#not-field}
+
+**The cause:** You're trying to promote a getter,
+but only instance fields can be promoted, not getters. 
+
+The compiler has no way to guarantee that a getter returns the same result every time.
+Because their stability can't be confirmed, getters are not safe to promote.
+
+**Example:**
+
+{:.bad}
+```dart
+abstract class C {
+  int? get _i; // CONTEXT
+}
+
+class Derived extends Base {
+  @override
+  final int? _i;
+  Derived(this._i);
+}
+
+f(Base b) {
+  if (b._i != null) {
+    int i = b._i; // ERROR
+  }
+}
+```
+
+**Message:**
+
+```nocode
+'_i' refers to a getter so it couldn’t be promoted.
+```
+
+**Solution:**
+
+Assign the getter to a local variable:
+
+{:.good}
+```dart
+f(Base b) {
+  final j = b._i;
+  if (i != null) {
+    int i = j; // OK
+  }
+}
+```
+
+#### Note about abstract fields
+
+If the getter in question is [abstract][], you might be able to enable field
+promotion and avoid the error by replacing the abstract getter with an [abstract field][].
+
+For example, this code produces an error because the getter `_i` can't be promoted
+and assigned to a non-nullable type:
+
+{:.bad}
+```dart
+abstract class Base {
+  int? get _i; // CONTEXT
+}
+
+class Derived extends Base {
+  @override
+  final int? _i;
+  Derived(this._i);
+}
+
+f(Base b) {
+  if (b._i != null) {
+    int i = b._i; // ERROR
```

Comment:
When we make the breaking change described in https://github.com/dart-lang/language/issues/3328#issuecomment-1792511446, this error will go away.

You might want to consider just dropping this subsection, since it's a subtlety that most people are not going to run into, and it will stop being a problem in a few months.

If you do decide to keep it, maybe mention that this is a known bug and that we're planning to fix it in a future release.

Reply:
Ok, I'll drop the section and add a temporary note.